googlemock/include/gmock/gmock-function-mocker.h

```diff
@@ -21,7 +21,7 @@
   GMOCK_INTERNAL_ASSERT_PARENTHESIS(_Spec);                                   \
   GMOCK_INTERNAL_ASSERT_VALID_SIGNATURE(                                      \
       GMOCK_PP_NARG0 _Args, GMOCK_INTERNAL_SIGNATURE(_Ret, _Args));           \
-  GMOCK_INTERNAL_ASSERT_VALID_SPEC(_Spec);                                    \
+  GMOCK_INTERNAL_ASSERT_VALID_SPEC(_Spec)                                     \
```

Comment:
`GMOCK_INTERNAL_ASSERT_VALID_SPEC` should not have a semicolon at the end of the definition IMHO.

Reply:
Removing the semicolon at the point of use was the “easy” way to get this to work. There’s also a “hard” way that involves duplicating some internal gtest macros in order to support this case, as distinct from other users of that macro machinery which are in different contexts.

Here’s what’s happening.

googlemock/include/gmock/gmock-function-mocker.h:64:
```
#define GMOCK_INTERNAL_ASSERT_VALID_SPEC(_Spec) \
  GMOCK_PP_FOR_EACH(GMOCK_INTERNAL_ASSERT_VALID_SPEC_ELEMENT, ~, _Spec)
```

googlemock/include/gmock/internal/gmock-pp.h:159:
```
#define GMOCK_PP_FOR_EACH(_Macro, _Data, _Tuple)                        \
  GMOCK_PP_CAT(GMOCK_PP_INTERNAL_FOR_EACH_IMPL_, GMOCK_PP_NARG0 _Tuple) \
  (0, _Macro, _Data, _Tuple)
```

For `GMOCK_PP_NARG0 _Tuple == 0`, googlemock/include/gmock/internal/gmock-pp.h:257:
```
#define GMOCK_PP_INTERNAL_FOR_EACH_IMPL_0(_i, _Macro, _Data, _Tuple)
```

So in that case, the entire `GMOCK_INTERNAL_ASSERT_VALID_SPEC` expands to nothing, and an extra dangling semicolon would be left behind.

For `GMOCK_PP_NARG0 _Tuple > 0`, the same problem would be experienced following the last element expansion. googlemock/include/gmock/internal/gmock-pp.h:258:
```
#define GMOCK_PP_INTERNAL_FOR_EACH_IMPL_1(_i, _Macro, _Data, _Tuple) \
  GMOCK_PP_INTERNAL_CALL_MACRO(_Macro, _i, _Data, GMOCK_PP_HEAD _Tuple)
#define GMOCK_PP_INTERNAL_FOR_EACH_IMPL_2(_i, _Macro, _Data, _Tuple)    \
  GMOCK_PP_INTERNAL_CALL_MACRO(_Macro, _i, _Data, GMOCK_PP_HEAD _Tuple) \
  GMOCK_PP_INTERNAL_FOR_EACH_IMPL_1(GMOCK_PP_INC(_i), _Macro, _Data,    \
                                    (GMOCK_PP_TAIL _Tuple))
```

`_Macro` here is `GMOCK_INTERNAL_ASSERT_VALID_SPEC_ELEMENT`, googlemock/include/gmock/gmock-function-mocker.h:122:
```
#define GMOCK_INTERNAL_ASSERT_VALID_SPEC_ELEMENT(_i, _, _elem)            \
  static_assert(                                                          \
      (GMOCK_PP_HAS_COMMA(GMOCK_INTERNAL_DETECT_CONST(_i, _, _elem)) +    \
       GMOCK_PP_HAS_COMMA(GMOCK_INTERNAL_DETECT_OVERRIDE(_i, _, _elem)) + \
       GMOCK_PP_HAS_COMMA(GMOCK_INTERNAL_DETECT_FINAL(_i, _, _elem)) +    \
       GMOCK_PP_HAS_COMMA(GMOCK_INTERNAL_DETECT_NOEXCEPT(_i, _, _elem)) + \
       GMOCK_INTERNAL_IS_CALLTYPE(_elem)) == 1,                           \
      GMOCK_PP_STRINGIZE(                                                 \
          _elem) " cannot be recognized as a valid specification modifier.");
```

There will always be a semicolon at the end of each element (from the text of `GMOCK_INTERNAL_ASSERT_VALID_SPEC_ELEMENT`), plus the trailing semicolon that followed `GMOCK_INTERNAL_ASSERT_VALID_SPEC` in the body of `GMOCK_INTERNAL_MOCK_METHOD_ARG_4`.

It sounds like you favor removing the reliance on the semicolon from the macro body (here, `GMOCK_INTERNAL_ASSERT_VALID_SPEC_ELEMENT`), and letting `GMOCK_PP_FOR_EACH` or a work-alike place them between expansions for each element, but not following the tail element. This is possible, but requires forking `GMOCK_PP_FOR_EACH` into another macro intended for use in class context. In between element expansions, it would insert a semicolon; for empty tuples, it’d place a harmless no-op that expects a semicolon terminator, such as `static_assert(true, "")`. `GMOCK_PP_FOR_EACH` itself can’t be modified because its other uses are in different contexts where these semicolons and static_asserts would be undesirable.

I’ve updated https://github.com/google/googletest/pull/2139 to this approach. Please let me know if you like it better.

+ @nico too.